File: quiz-client/src/types/quiz.ts
export interface QuizSession {
  id: string;
  quiz_id: string;
  status: 'waiting' | 'active' | 'completed';
  current_question: number;
  questions: Question[];
  participants: string[];
  start_time: string;
  end_time?: string;
}

export interface SessionStateMessage {
  type: 'session_state';
  session_id: string;
  status: 'waiting' | 'active' | 'completed';
  current_question: number;
  questions: Question[];
  participants: string[];
  quiz_id: string;
  start_time: string;
}

export interface Question {
  id: string;
  text: string;
  type: 'multiple_choice' | 'true_false';
  options: string[];
  correct_answer: string;
  points: number;
  time_limit: number;
}

export interface LeaderboardEntry {
  user_id: string;
  score: number;
}

export interface Answer {
  session_id: string;
  question_id: string;
  user_id: string;
  answer: string;
  timestamp?: string;
}

export interface WebSocketMessage {
  type: string;
  [key: string]: any;
}

export interface QuizApiError {
  status: string;
  message: string;
  status_code: number;
}

File: quiz-service/app/models/quiz.py
from datetime import datetime
from enum import Enum
from typing import List, Optional

from pydantic import BaseModel, Field


class QuestionType(str, Enum):
    MULTIPLE_CHOICE = "multiple_choice"
    TRUE_FALSE = "true_false"


class Question(BaseModel):
    id: str = Field(..., description="Question ID")
    text: str
    type: QuestionType
    options: List[str]
    correct_answer: str
    points: int = 1
    time_limit: int = 30  # seconds


class Quiz(BaseModel):
    id: Optional[str] = Field(None, alias="_id")
    title: str
    description: str
    questions: List[Question]
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)


class QuizSession(BaseModel):
    id: str
    quiz_id: str
    status: str = "waiting"  # "waiting", "active", "completed"
    current_question: int = 0
    questions: List[Question]
    start_time: datetime
    end_time: Optional[datetime] = None
    participants: List[str] = []
    updated_at: Optional[datetime] = None

    class Config:
        json_encoders = {datetime: lambda v: v.isoformat()}


class Answer(BaseModel):
    session_id: str
    question_id: str
    user_id: str
    answer: str
    timestamp: datetime = Field(default_factory=datetime.utcnow)
